Check for type collisions when registering a free parameter against existing inputs

Prototype branch started here to explore fixing this, which always allow input parameters to be declared at the time of registering them: https://github.com/amazon-braket/autoqasm/tree/rmshaffer/input-param-name-conflict

But as discussed above, this results in a failure in `test_pulse_freeparameter_condition` due to the name collision between the input parameter and the free parameter:
```
RuntimeError: Program has conflicting variables with name duration

C:\Miniconda3\envs\autoqasm-py39\lib\site-packages\oqpy\program.py:203: RuntimeError

E           RuntimeError: in user code:
E
E               File "C:\Repos\autoqasm\test\unit_tests\autoqasm\test_pulse.py", line 235, in my_program  *
E                   if duration > duration2:
E               File "C:\Repos\autoqasm\src\autoqasm\operators\comparisons.py", line 90, in gt_
E                   return _aq_gt(a, b)
E               File "C:\Repos\autoqasm\src\autoqasm\operators\comparisons.py", line 101, in _aq_gt
E                   oqpy_program.set(result, a > b)
E               File "C:\Miniconda3\envs\autoqasm-py39\lib\site-packages\oqpy\program.py", line 667, in set
E                   self._do_assignment(var, "=", value)
E               File "C:\Miniconda3\envs\autoqasm-py39\lib\site-packages\oqpy\program.py", line 652, in _do_assignment
E                   to_ast(self, value),
E               File "C:\Miniconda3\envs\autoqasm-py39\lib\site-packages\oqpy\base.py", line 529, in to_ast
E                   return item.to_ast(program)
E               File "C:\Miniconda3\envs\autoqasm-py39\lib\site-packages\oqpy\base.py", line 417, in to_ast
E                   return ast.BinaryExpression(self.op, to_ast(program, self.lhs), to_ast(program, self.rhs))
E               File "C:\Miniconda3\envs\autoqasm-py39\lib\site-packages\oqpy\base.py", line 529, in to_ast
E                   return item.to_ast(program)
E               File "C:\Miniconda3\envs\autoqasm-py39\lib\site-packages\oqpy\classical_types.py", line 218, in to_ast
E                   program._add_var(self)
E               File "C:\Miniconda3\envs\autoqasm-py39\lib\site-packages\oqpy\program.py", line 203, in _add_var
E                   raise RuntimeError(f"Program has conflicting variables with name {name}")
```

--- a/src/autoqasm/program/program.py
+++ b/src/autoqasm/program/program.py
@@ -483,23 +483,36 @@ class ProgramConversionContext:
 
         Raises:
             NotImplementedError: If the parameter type is not supported.
+            errors.ParameterTypeError: If ``name`` is already registered with
+                a different type.
         """
-        # TODO (#814): add type validation against existing inputs
-        if name not in self._input_parameters:
-            aq_type = aq_types.map_parameter_type(param_type)
-            if aq_type not in [oqpy.FloatVar, oqpy.IntVar, oqpy.BoolVar]:
-                raise NotImplementedError(param_type)
+        if name in self._input_parameters:
+            existing = self._input_parameters[name]
+            expected_type = aq_types.map_parameter_type(param_type)
+            if type(existing) is not expected_type:
+                raise errors.ParameterTypeError(
+                    f'Input parameter "{name}" was already registered as '
+                    f"{type(existing).__name__}, but a later use implies "
+                    f"{expected_type.__name__}. Make the types match, for "
+                    "example by updating the @aq.main signature to match "
+                    "the type expected by the gate or subroutine that "
+                    "consumes the parameter."
+                )
+            return
+        aq_type = aq_types.map_parameter_type(param_type)
+        if aq_type not in [oqpy.FloatVar, oqpy.IntVar, oqpy.BoolVar]:
+            raise NotImplementedError(param_type)
 
-            # In case a FreeParameter has already created a FloatVar somewhere else,
-            # we use need_declaration=False to avoid OQPy raising name conflict errors.
-            if aq_type == oqpy.FloatVar:
-                var = aq_type("input", name=name, needs_declaration=False)
-                var.size = None
-                var.type.size = None
-            else:
-                var = aq_type("input", name=name)
-            self._input_parameters[name] = var
-            return var
+        # In case a FreeParameter has already created a FloatVar somewhere else,
+        # we use need_declaration=False to avoid OQPy raising name conflict errors.
+        if aq_type == oqpy.FloatVar:
+            var = aq_type("input", name=name, needs_declaration=False)
+            var.size = None
+            var.type.size = None
+        else:
+            var = aq_type("input", name=name)
+        self._input_parameters[name] = var
+        return var
 
     def register_output_parameter(
         self, name: str, value: bool | float | oqpy.base.Var | oqpy.OQPyExpression | None

--- a/test/unit_tests/autoqasm/test_parameter_type_collision.py
+++ b/test/unit_tests/autoqasm/test_parameter_type_collision.py
@@ -1,0 +1,84 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+"""Tests for type-collision detection on input-parameter registration."""
+
+import pytest
+from braket.circuits.free_parameter import FreeParameter
+
+import autoqasm as aq
+from autoqasm.instructions import rx
+
+
+def test_signature_int_conflicts_with_free_parameter_float() -> None:
+    """A signature annotation of ``int`` conflicts with the implicit
+    ``float`` type produced by passing ``FreeParameter(...)`` to a gate
+    angle. We must raise rather than silently coerce one type to the other.
+    """
+
+    @aq.main
+    def collision(theta: int):
+        rx(0, FreeParameter("theta"))
+
+    with pytest.raises(aq.errors.ParameterTypeError, match="theta"):
+        collision.build()
+
+
+def test_signature_bool_conflicts_with_free_parameter_float() -> None:
+    @aq.main
+    def collision(theta: bool):
+        rx(0, FreeParameter("theta"))
+
+    with pytest.raises(aq.errors.ParameterTypeError, match="theta"):
+        collision.build()
+
+
+def test_matching_types_are_ok() -> None:
+    """Matching types (float + float) must not raise."""
+
+    @aq.main
+    def ok(theta: float):
+        rx(0, FreeParameter("theta"))
+
+    ir = ok.build().to_ir()
+    assert "input float theta;" in ir
+
+
+def test_bare_free_parameter_is_fine() -> None:
+    """Without a conflicting signature annotation, ``FreeParameter`` keeps
+    working as it always has."""
+
+    @aq.main
+    def ok():
+        rx(0, FreeParameter("theta"))
+
+    ir = ok.build().to_ir()
+    assert "input float theta;" in ir
+
+
+def test_register_input_parameter_same_type_is_idempotent() -> None:
+    """Registering the same parameter twice with the same type is a no-op."""
+    with aq.build_program() as ctx:
+        first = ctx.register_input_parameter("theta", float)
+        result = ctx.register_input_parameter("theta", float)
+        # The second call returns None (already registered) and the
+        # underlying var object is unchanged.
+        assert result is None
+        assert ctx._input_parameters["theta"] is first
+
+
+def test_register_input_parameter_conflicting_types_raises() -> None:
+    with aq.build_program() as ctx:
+        ctx.register_input_parameter("theta", int)
+        with pytest.raises(aq.errors.ParameterTypeError, match="theta"):
+            ctx.register_input_parameter("theta", float)

